File: samples/AdventureWorksLT.VB/SalesOrderDetail.vb
<Computation(SalesOrderDetail._ComputeLineTotal)>
<CheckConstraint(SalesOrderDetail._CK_SalesOrderDetail_OrderQty, GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderDetail_OrderQty), Description:="Check constraint [OrderQty] > (0)")>
<CheckConstraint(SalesOrderDetail._CK_SalesOrderDetail_UnitPrice, GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderDetail_UnitPrice), Description:="heck constraint [UnitPrice] >= (0.00)")>
<CheckConstraint(SalesOrderDetail._CK_SalesOrderDetail_UnitPriceDiscount, GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderDetail_UnitPriceDiscount), Description:="Check constraint [UnitPriceDiscount] >= (0.00)")>
<DbIndex(SalesOrderDetail._IX_SalesOrderDetail_ProductID, Description:="Nonclustered index.")>
Public Class SalesOrderDetail
    Inherits BaseModel(Of PK)

    <DbPrimaryKey("PK_SalesOrderDetail_SalesOrderID_SalesOrderDetailID", Description:="Clustered index created by a primary key constraint.")>
    Public NotInheritable Class PK
        Inherits CandidateKey

        Public Sub New(salesOrderID As _Int32, salesOrderDetailID As _Int32)
            MyBase.New(salesOrderID, salesOrderDetailID)
        End Sub
    End Class

    Public Class Key
        Inherits Key(Of PK)

        Shared Sub New()
            Register(Function(x As Key) x.SalesOrderID, _SalesOrderID)
            Register(Function(x As Key) x.SalesOrderDetailID, _SalesOrderDetailID)
        End Sub

        Protected Overrides Function CreatePrimaryKey() As PK
            Return New PK(SalesOrderID, SalesOrderDetailID)
        End Function

        Private m_SalesOrderID As _Int32
        Public Property SalesOrderID As _Int32
            Get
                Return m_SalesOrderID
            End Get
            Private Set
                m_SalesOrderID = Value
            End Set
        End Property

        Private m_SalesOrderDetailID As _Int32
        Public Property SalesOrderDetailID As _Int32
            Get
                Return m_SalesOrderDetailID
            End Get
            Private Set
                m_SalesOrderDetailID = Value
            End Set
        End Property
    End Class

    Public Shared ReadOnly _SalesOrderID As Mounter(Of _Int32) = RegisterColumn(Function(x As SalesOrderDetail) x.SalesOrderID)
    Public Shared ReadOnly _SalesOrderDetailID As Mounter(Of _Int32) = RegisterColumn(Function(x As SalesOrderDetail) x.SalesOrderDetailID)
    Public Shared ReadOnly _OrderQty As Mounter(Of _Int16) = RegisterColumn(Function(x As SalesOrderDetail) x.OrderQty)
    Public Shared ReadOnly _ProductID As Mounter(Of _Int32) = RegisterColumn(Function(x As SalesOrderDetail) x.ProductID)
    Public Shared ReadOnly _UnitPrice As Mounter(Of _Decimal) = RegisterColumn(Function(x As SalesOrderDetail) x.UnitPrice)
    Public Shared ReadOnly _UnitPriceDiscount As Mounter(Of _Decimal) = RegisterColumn(Function(x As SalesOrderDetail) x.UnitPriceDiscount)
    Public Shared ReadOnly _LineTotal As Mounter(Of _Decimal) = RegisterColumn(Function(x As SalesOrderDetail) x.LineTotal)

    Protected NotOverridable Overrides Function CreatePrimaryKey() As PK
        Return New PK(SalesOrderID, SalesOrderDetailID)
    End Function

    Private m_FK_SalesOrderHeader As SalesOrderHeader.PK
    Public ReadOnly Property FK_SalesOrderHeader As SalesOrderHeader.PK
        Get
            If m_FK_SalesOrderHeader Is Nothing Then
                m_FK_SalesOrderHeader = New SalesOrderHeader.PK(SalesOrderID)
            End If
            Return m_FK_SalesOrderHeader
        End Get
    End Property

    Private m_FK_Product As Product.PK
    Public ReadOnly Property FK_Product As Product.PK
        Get
            If m_FK_Product Is Nothing Then
                m_FK_Product = New Product.PK(ProductID)
            End If
            Return m_FK_Product
        End Get
    End Property

    Private m_SalesOrderID As _Int32
    <DbColumn(Description:="Primary key. Foreign key to SalesOrderHeader.SalesOrderID.")>
    Public Property SalesOrderID As _Int32
        Get
            Return m_SalesOrderID
        End Get
        Private Set
            m_SalesOrderID = Value
        End Set
    End Property

    Private m_SalesOrderDetailID As _Int32
    <Identity(1, 1)>
    <DbColumn(Description:="Primary key. One incremental unique number per product sold.")>
    Public Property SalesOrderDetailID As _Int32
        Get
            Return m_SalesOrderDetailID
        End Get
        Private Set
            m_SalesOrderDetailID = Value
        End Set
    End Property

    Private m_OrderQty As _Int16
    <Required>
    <DbColumn(Description:="Quantity ordered per product.")>
    Public Property OrderQty As _Int16
        Get
            Return m_OrderQty
        End Get
        Private Set
            m_OrderQty = Value
        End Set
    End Property

    Private m_ProductID As _Int32
    <Required>
    <DbColumn(Description:="Product sold to customer. Foreign key to Product.ProductID.")>
    Public Property ProductID As _Int32
        Get
            Return m_ProductID
        End Get
        Private Set
            m_ProductID = Value
        End Set
    End Property

    Private m_UnitPrice As _Decimal
    <Required>
    <SqlMoney>
    <DbColumn(Description:="Selling price of a single product.")>
    Public Property UnitPrice As _Decimal
        Get
            Return m_UnitPrice
        End Get
        Private Set
            m_UnitPrice = Value
        End Set
    End Property

    Private m_UnitPriceDiscount As _Decimal
    <Required>
    <SqlMoney>
    <DefaultValue(GetType(Decimal), "0", Name:="DF_SalesOrderDetail_UnitPriceDiscount")>
    <DbColumn(Description:="Discount amount.")>
    Public Property UnitPriceDiscount As _Decimal
        Get
            Return m_UnitPriceDiscount
        End Get
        Private Set
            m_UnitPriceDiscount = Value
        End Set
    End Property

    Private m_LineTotal As _Decimal
    <Required>
    <SqlMoney>
    <DbColumn(Description:="Per product subtotal. Computed as UnitPrice * (1 - UnitPriceDiscount) * OrderQty.")>
    Public Property LineTotal As _Decimal
        Get
            Return m_LineTotal
        End Get
        Private Set
            m_LineTotal = Value
        End Set
    End Property

    Friend Const _ComputeLineTotal = NameOf(ComputeLineTotal)
    <_Computation>
    Private Sub ComputeLineTotal()
        LineTotal.ComputedAs((UnitPrice * (_Decimal.[Const](1) - UnitPriceDiscount) * OrderQty).IfNull(_Decimal.[Const](0)))
    End Sub

    Friend Const _CK_SalesOrderDetail_OrderQty = NameOf(CK_SalesOrderDetail_OrderQty)
    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderDetail_OrderQty As _Boolean
        Get
            Return OrderQty > _Decimal.Const(0)
        End Get
    End Property

    Friend Const _CK_SalesOrderDetail_UnitPrice = NameOf(CK_SalesOrderDetail_UnitPrice)
    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderDetail_UnitPrice As _Boolean
        Get
            Return UnitPrice >= _Decimal.Const(0)
        End Get
    End Property

    Friend Const _CK_SalesOrderDetail_UnitPriceDiscount = NameOf(CK_SalesOrderDetail_UnitPriceDiscount)
    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderDetail_UnitPriceDiscount As _Boolean
        Get
            Return UnitPriceDiscount >= _Decimal.Const(0)
        End Get
    End Property

    Friend Const _IX_SalesOrderDetail_ProductID = NameOf(IX_SalesOrderDetail_ProductID)
    <_DbIndex>
    Private ReadOnly Property IX_SalesOrderDetail_ProductID As ColumnSort()
        Get
            Return New ColumnSort() {ProductID}
        End Get
    End Property
End Class

File: samples/AdventureWorksLT.Data.SqlServer/AdventureWorksLT.Data/SalesOrderDetail.cs
using DevZest.Data;
using DevZest.Data.Annotations;
using DevZest.Data.SqlServer;

namespace DevZest.Samples.AdventureWorksLT
{
    [Computation(nameof(ComputeLineTotal))]
    [CheckConstraint(nameof(CK_SalesOrderDetail_OrderQty), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderDetail_OrderQty), Description = "Check constraint [OrderQty] > (0)")]
    [CheckConstraint(nameof(CK_SalesOrderDetail_UnitPrice), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderDetail_UnitPrice), Description = "Check constraint [UnitPrice] >= (0.00)")]
    [CheckConstraint(nameof(CK_SalesOrderDetail_UnitPriceDiscount), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderDetail_UnitPriceDiscount), Description = "Check constraint [UnitPriceDiscount] >= (0.00)")]
    [DbIndex(nameof(IX_SalesOrderDetail_ProductID), Description = "Nonclustered index.")]
    public class SalesOrderDetail : BaseModel<SalesOrderDetail.PK>
    {
        [DbPrimaryKey("PK_SalesOrderDetail_SalesOrderID_SalesOrderDetailID", Description = "Clustered index created by a primary key constraint.")]
        public sealed class PK : CandidateKey
        {
            public PK(_Int32 salesOrderID, _Int32 salesOrderDetailID)
                : base(salesOrderID, salesOrderDetailID)
            {
            }
        }

        public class Key : Key<PK>
        {
            static Key()
            {
                Register((Key _) => _.SalesOrderID, _SalesOrderID);
                Register((Key _) => _.SalesOrderDetailID, _SalesOrderDetailID);
            }

            protected override PK CreatePrimaryKey()
            {
                return new PK(SalesOrderID, SalesOrderDetailID);
            }

            public _Int32 SalesOrderID { get; private set; }

            public _Int32 SalesOrderDetailID { get; private set; }
        }

        public static readonly Mounter<_Int32> _SalesOrderID = RegisterColumn((SalesOrderDetail _) => _.SalesOrderID);
        public static readonly Mounter<_Int32> _SalesOrderDetailID = RegisterColumn((SalesOrderDetail _) => _.SalesOrderDetailID);
        public static readonly Mounter<_Int16> _OrderQty = RegisterColumn((SalesOrderDetail _) => _.OrderQty);
        public static readonly Mounter<_Int32> _ProductID = RegisterColumn((SalesOrderDetail _) => _.ProductID);
        public static readonly Mounter<_Decimal> _UnitPrice = RegisterColumn((SalesOrderDetail _) => _.UnitPrice);
        public static readonly Mounter<_Decimal> _UnitPriceDiscount = RegisterColumn((SalesOrderDetail _) => _.UnitPriceDiscount);
        public static readonly Mounter<_Decimal> _LineTotal = RegisterColumn((SalesOrderDetail _) => _.LineTotal);

        public SalesOrderDetail()
        {
        }

        protected sealed override PK CreatePrimaryKey()
        {
            return new PK(SalesOrderID, SalesOrderDetailID);
        }

        private SalesOrderHeader.PK _fk_salesOrderHeader;
        public SalesOrderHeader.PK FK_SalesOrderHeader
        {
            get { return _fk_salesOrderHeader ?? (_fk_salesOrderHeader = new SalesOrderHeader.PK(SalesOrderID)); }
        }

        private Product.PK _fk_product;
        public Product.PK FK_Product
        {
            get { return _fk_product ?? (_fk_product = new Product.PK(ProductID)); }
        }

        [DbColumn(Description = "Primary key. Foreign key to SalesOrderHeader.SalesOrderID.")]
        public _Int32 SalesOrderID { get; private set; }

        [Identity]
        [DbColumn(Description = "Primary key. One incremental unique number per product sold.")]
        public _Int32 SalesOrderDetailID { get; private set; }

        [Required]
        [DbColumn(Description = "Quantity ordered per product.")]
        public _Int16 OrderQty { get; private set; }

        [Required]
        [DbColumn(Description = "Product sold to customer. Foreign key to Product.ProductID.")]
        public _Int32 ProductID { get; private set; }

        [Required]
        [SqlMoney]
        [DbColumn(Description = "Selling price of a single product.")]
        public _Decimal UnitPrice { get; private set; }

        [Required]
        [SqlMoney]
        [DefaultValue(typeof(decimal), "0", Name = "DF_SalesOrderDetail_UnitPriceDiscount")]
        [DbColumn(Description = "Discount amount.")]
        public _Decimal UnitPriceDiscount { get; private set; }

        [Required]
        [SqlMoney]
        [DbColumn(Description = "Per product subtotal. Computed as UnitPrice * (1 - UnitPriceDiscount) * OrderQty.")]
        public _Decimal LineTotal { get; private set; }

        [_Computation]
        private void ComputeLineTotal()
        {
            LineTotal.ComputedAs((UnitPrice * (_Decimal.Const(1) - UnitPriceDiscount) * OrderQty).IfNull(_Decimal.Const(0)));
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderDetail_OrderQty
        {
            get { return OrderQty > _Decimal.Const(0); }
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderDetail_UnitPrice
        {
            get { return UnitPrice >= _Decimal.Const(0); }
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderDetail_UnitPriceDiscount
        {
            get { return UnitPriceDiscount >= _Decimal.Const(0); }
        }

        [_DbIndex]
        private ColumnSort[] IX_SalesOrderDetail_ProductID => new ColumnSort[] { ProductID };
    }
}
